internal/controller/object/object.go

```diff
@@ -302,12 +350,19 @@ func (c *external) Create(ctx context.Context, mg resource.Managed) (managed.Ext
 		return managed.ExternalCreation{}, err
 	}
 
+	if c.ssaExtractor != nil {
```

Comment:
I don't think this will make things better since there will be lots of duplication in that case. 

While it is not perfect because of the if branches in its current shape, I found it more readable since it only impacts where we calculate the desired state and do the applies.

Reply:
I have two concerns here:

* Adding branches to almost every method of the ExternalClient that toggle its behavior.
* The condition of the branch - the presence of an `ssaExtractor`.

I feel two less complex implementation is better than one more complex implementation that has two different modes of operation. Each less complex implementation is easier to reason about, test, etc in isolation. Looking at this `ExternalClient`, it seems like `Observe` is the only method with much complexity right now - the rest I wouldn't be concerned about duplicating.

Using `ssaExtractor` to determine whether to SSA or works, but you're using a side effect of SSA being enabled in one place to determine whether SSA is enabled in another place. The nilness of the extractor is being used not only to gate whether extraction happens, but whether SSA happens at all.

If this was c/c I'd push back harder on this direction, but give I'm not a maintainer on this repo I'm okay to disagree and go with this more pragmatic approach.

Reply:
I still don't like having multiple external client implementations due to duplication, but I see your point.

Just pushed a [commit](https://github.com/crossplane-contrib/provider-kubernetes/pull/246/commits/c6d5a72468ae4afc0a3568e2349465f958888ec8) hopefully resolving both concerns, basically by having two different implementations of a new interface, one for old patch-based and another for SSA. This way, I could avoid branches without duplication.

Reply:
I still need to fix unit tests ...

Reply:
Thanks, the new approach LGTM.